clar2wasm/tests/wasm-generation/principal.rs

```diff
@@ -0,0 +1,230 @@
+                &hash_bytes.clone().expect_buff(20).unwrap(),
+                contract.as_deref()
+            );
+
+            let data = contract.map(|ctc| Box::new(
+                Value::Sequence(SequenceData::String(CharType::ASCII(ASCIIData {
+                    data: ctc.into_bytes()
+                })))
+            ));
+
+            let expected = match version_byte {
+                // Since tests runs on a Testnet version,
+                // version_bytes single_sig (0x1A) || multi_sig (0x15), for Testnet,
+                // will return an Ok value.
+                0x1A | 0x15 => Value::okay(
+                    TupleData::from_data(vec![
+                        ("hash-bytes".into(), hash_bytes),
+                        ("name".into(), Value::Optional(OptionalData { data })),
+                        (
+                            "version".into(),
+                            Value::Sequence(SequenceData::Buffer(BuffData {
+                                data: vec![version_byte as u8],
+                            })),
+                        ),
+                    ])
+                    .unwrap()
+                    .into()
+                ),
+                0x00..=0x1F => Ok(create_error_destruct(hash_bytes, version_byte as u8, data)),
+                _ => Ok(create_error_destruct(hash_bytes, version_byte as u8, data)),
+            }.unwrap();
+
+            crosscheck(
+                &format!("(principal-destruct? {})", PropValue::from(expected_principal.clone())),
+                Ok(Some(expected)),
+            );
+        }
+    }
+
+    proptest! {
+        #![proptest_config(runtime_config())]
+
+        #[test]
+        fn crosscheck_is_standard(
+            principal in standard_principal().prop_map(PropValue::from)
+        ) {
+            let principal_str = principal.to_string();
+            let expected_in_testnet = matches!(principal_str.get(0..3), Some("'ST") | Some("'SN"));
+            let expected_in_mainnet = matches!(principal_str.get(0..3), Some("'SP") | Some("'SM"));
+
+            let crosscheck_for = |network: &Network, expected: bool| {
+                crosscheck_with_network(
+                    network,
+                    &format!("(is-standard {})", principal),
+                    Ok(Some(Value::Bool(expected))),
+                );
+            };
+
+            crosscheck_for(&Network::Mainnet, expected_in_mainnet);
+            crosscheck_for(&Network::Testnet, expected_in_testnet);
+        }
+    }
+
+    proptest! {
+        #![proptest_config(runtime_config())]
+
+        #[test]
+        fn crosscheck_principal_of(
+            seed in proptest::collection::vec(any::<u8>(), 20)
+        ) {
+
+            let private_key = Secp256k1PrivateKey::from_seed(&seed);
+            let public_key = Secp256k1PublicKey::from_private(&private_key);
+
+            let snippet = &format!("(is-standard (try! (principal-of? 0x{})))", public_key.to_hex());
```

Comment:
Why is this snippet so complex? Can't we just call `principal-of?` and check the result?

Reply:
Isn't that complex, is it? :-) Changed anyway.

Reply:
Well, yes it is. Let's imagine this snippet fails. What would be the cause of it? The call to `principal-of?`, the call to `try!`, or the call to `is-standard`?

What if `principal-of` gives a random bad answer, but `is-standard` is bugged, and the test succeeds ?